runtime/compare.c

```diff
@@ -280,9 +283,10 @@ static intnat do_compare_val(struct compare_stack* stk,
       if (sz1 > 1) {
         sp++;
         if (sp >= stk->limit) sp = compare_resize_stack(stk, sp);
-        sp->v1 = &Field(v1, 1);
-        sp->v2 = &Field(v2, 1);
-        sp->count = sz1 - 1;
+        sp->v1 = v1;
+        sp->v2 = v2;
+        sp->offset = Val_long(1);
+        sp->size = Val_long(sz1);
       }
       /* Continue comparison with first field */
```

Comment:
Recall that the `continue;` after this needs to be a jump to the poll block to ensure that the comparison in e.g.
```ocaml
type nat = S of nat | Z;;
let rec inf = S inf;;
inf = inf;;
```
polls.

Reply:
It looks like you've moved it anyway.

Reply:
I moved the poll check to the entry of the `while` loop as an easy way to fix this issue. I haven't tried (yet) to integrate your code to run regularly inside loops, because I find it a bit too complex to my taste.